libvmi/driver/xen/xen.c

```diff
@@ -788,6 +788,27 @@ xen_get_memsize(
     return VMI_SUCCESS;
 }
 
+status_t
+xen_request_page_fault(
+    vmi_instance_t vmi,
+    unsigned long vcpu,
+    uint64_t virtual_address,
+    uint32_t error_code)
+{
+
+    xen_instance_t *xen = xen_get_instance(vmi);
+    int rc = xen->libxcw.xc_hvm_inject_trap(xen->xchandle, xen->domainid,
```

Comment:
Please add a safety check (behind the ifdef) that grabs cs_sel and checks the CPL of the vCPU to verify it's safe to call this function.

Reply:
Do you mean ss_sel? It looks like then xen check is done with ss_sel 
  unsigned int vmx_get_cpl(void)
  {
      unsigned long attr;
  
      __vmread(GUEST_SS_AR_BYTES, &attr);
  
      return (attr >> 5) & 3;
  }
I guess that the guest_mode check is  if((ss_attr >> 5) == 3)
Is this what you want?

Reply:
That's actually the `RPL` you reference. In this context we should check the `CPL`. See https://css.csail.mit.edu/6.858/2014/readings/i386/s06_03.htm

Reply:
Are we discussing the use of CS vs. SS? If so, the Intel SDM says in "5.5 PRIVILEGE LEVELS":
```
Current privilege level (CPL) — The CPL is the privilege level of the currently executing program or task. It is stored in bits 0 and 1 of the CS and SS segment registers. Normally, the CPL is equal to the privilege level of the code segment from which instructions are being fetched. The processor changes the CPL when program control is transferred to a code segment with a different privilege level. The CPL is treated slightly differently when accessing conforming code segments. Conforming code segments can be accessed from any privilege level that is equal to or numerically greater (less privileged) than the DPL of the conforming code segment. Also, the CPL is not changed when the processor accesses a conforming code segment that has a different privilege level than the CPL.
```
https://software.intel.com/sites/default/files/managed/a4/60/325384-sdm-vol-3abcd.pdf
Also, Jan's been quite adamant about using SS:
https://lists.xenproject.org/archives/html/xen-devel/2014-07/msg00415.html
And the KVM developers have switched from CS to SS (if I'm reading their patch correctly):
https://lore.kernel.org/patchwork/patch/464719/

Reply:
Yay for x86 being straight forward :) AFAIU cs_sel and the ss attr CPL bits should be equal for most cases, but in corner-cases it's safer to use ss attr bits. TIL